src/IndexingMsg.chpl

```diff
@@ -380,9 +380,147 @@ module IndexingMsg
+                return new MsgTuple(errorMsg,MsgType.ERROR);
+            }
+        }
+
+        proc doMultiIdx(type arrType, type idxType): MsgTuple throws {
+            const eIn = toSymEntry(gEnt, arrType, nd),
+                  eIdxs = for g in gIdx do toSymEntry(g, idxType, 1),
+                  (valid, outRankIdx, outShape) = multiIndexShape(eIn.tupShape, idxDims, outSize);
+
+            if valid {
+                var eOut = st.addEntry(rname, (...outShape), arrType);
+
+                forall i in eOut.a.domain with (
+                    var agg = newSrcAggregator(arrType),
+                    in idxDims // short array: create a per-task copy
+                ) {
+                    const idx = if nd == 1 then (i, ) else i,
+                          iIdxArray = idx[outRankIdx];
+
+                    var inIdx = if nd == 1 then (i, ) else i;
+
+                    for (rank, j) in zip(idxDims, 0..<nIdxArrays) do
+                        inIdx[rank] = eIdxs[j].a[iIdxArray]:int;
+
+                    agg.copy(eOut.a[idx], eIn.a[inIdx]);
+                }
+
+                const repMsg =  "created " + st.attrib(rname);
+                imLogger.debug(getModuleName(),pn,getLineNumber(),repMsg);
+                return new MsgTuple(repMsg, MsgType.NORMAL);
+            } else {
+                const errorMsg = "Invalid index dimensions: %? for %iD array".doFormat(idxDims, nd);
+                imLogger.error(getModuleName(),pn,getLineNumber(),errorMsg);
+                return new MsgTuple(errorMsg,MsgType.ERROR);
             }
         }
+
+        select (gEnt.dtype, idxDT) {
+            when (DType.Int64, DType.Int64) do return doMultiIdx(int, int);
+            when (DType.Int64, DType.UInt64) do return doMultiIdx(int, uint);
+            when (DType.UInt64, DType.Int64) do return doMultiIdx(uint, int);
+            when (DType.UInt64, DType.UInt64) do return doMultiIdx(uint, uint);
+            when (DType.Float64, DType.Int64) do return doMultiIdx(real, int);
+            when (DType.Float64, DType.UInt64) do return doMultiIdx(real, uint);
+            when (DType.Bool, DType.Int64) do return doMultiIdx(bool, int);
+            when (DType.Bool, DType.UInt64) do return doMultiIdx(bool, uint);
+            when (DType.BigInt, DType.Int64) do return doMultiIdx(bigint, int);
+            when (DType.BigInt, DType.UInt64) do return doMultiIdx(bigint, uint);
+            otherwise {
+                var errorMsg = notImplementedError(pn,dtype2str(gEnt.dtype));
+                imLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);
+                return new MsgTuple(errorMsg,MsgType.ERROR);
+            }
+        }
+    }
+
+    private proc multiIndexShape(inShape: ?N*int, idxDims: [?d] int, outSize: int): (bool, int, N*int) {
+        var minShape: N*int = inShape,
+            firstRank = -1;
+
+        if d.size > N then return (false, -1, minShape);
```

Comment:
It seems like this `-1` is conveying the same information as `false`. I don't know if we need the additional `bool` value, but I also don't think it's hurting anything, so don't need to change. Am I missing something here though?

Reply:
There are cases where the second value in the tuple can be greater than `-1`, but the first value is still `false`. For example, if one of the values in `idxDims` is out of bounds, but `firstRank` has already been set, the function could return `(false, 1, ...)`.

This did make me realize that I should be putting the variable `firstRank` in that tuple here instead of the hard-coded `-1`, so I fixed that.